testhelpers/docker_registry.go

```diff
@@ -96,11 +94,13 @@ func ReadOnly(handler http.Handler) http.Handler {
 	})
 }
 
-func Custom(handler http.Handler, permissions map[string]ImagePrivileges) http.Handler {
+func Custom(basicAuthHandler http.Handler, regHandler http.Handler, permissions map[string]ImagePrivileges) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		if permission, ok := permissions[extractImageName(r.URL.Path)]; ok {
 			if r.Method == "GET" || r.Method == "HEAD" {
-				if !permission.readable {
+				if permission.readable {
+					regHandler.ServeHTTP(w, r)
```

Comment:
Wouldn't we also want to "wave through" a put request where the image has been explicitly configured to be writeable? Should we require basic auth in that case (I'd argue no). So there would be a similar extra lines right around line 109...

Reply:
If the image is writable, it will delegate the request to the Basic Auth handler that will check if you sent the credentials in the request, like the following image:

<img width="609" alt="registry-3" src="https://user-images.githubusercontent.com/1181799/129104983-87f968a4-92ae-482c-9bd7-7e5cf8a27d7c.png">

Reply:
The idea of all this stuff, was because to implement the read-only behavior without forcing the client to send the Basic Auth value in the request, but the Write operation must validate the Auth, right? or you mean if the image is writable (by anyone) we can skip the validation? Not sure if we can do that in the registry right now, in that case, yes! we can add a similar logic in line 109

Reply:
That's interesting - I think you could make the case for it working either way. It's probably worth a quick discussion!

Reply:
I think we might need a `return` after we call `regHandler.ServeHTTP()`. What do you think of the following?

```
var (
			image ImagePrivileges
			ok    bool
		)
		if image, ok = permissions[extractImageName(r.URL.Path)]; !ok {
			basicAuthHandler.ServeHTTP(w, r)
			return
		}
		if r.Method == "GET" || r.Method == "HEAD" {
			if !image.readable {
				w.WriteHeader(401)
				_, _ = w.Write([]byte("Unauthorized.\n"))
				return
			}
		} else { // assume write request
			if !image.writable {
				w.WriteHeader(401)
				_, _ = w.Write([]byte("Unauthorized.\n"))
				return
			}
		}
		regHandler.ServeHTTP(w, r)
```